test/benchmark_utils/test_benchmark_utils.py

```diff
@@ -0,0 +1,829 @@
+    )
+
+    return stats_no_data, stats_with_data
+
+
+class TestBenchmarkUtils(TestCase):
+    def test_timer(self):
+        timer = benchmark_utils.Timer(
+            stmt="torch.ones(())",
+        )
+        sample = timer.timeit(5).median
+        self.assertIsInstance(sample, float)
+
+        median = timer.blocked_autorange(min_run_time=0.01).median
+        self.assertIsInstance(median, float)
+
+        # We set a very high threshold to avoid flakiness in CI.
+        # The internal algorithm is tested in `test_adaptive_timer`
+        median = timer.adaptive_autorange(threshold=0.5).median
+
+        # Test that multi-line statements work properly.
+        median = benchmark_utils.Timer(
+            stmt="""
+                with torch.no_grad():
+                    y = x + 1""",
+            setup="""
+                x = torch.ones((1,), requires_grad=True)
+                for _ in range(5):
+                    x = x + 1.0""",
+        ).timeit(5).median
+        self.assertIsInstance(sample, float)
+
+    class _MockTimer:
+        _seed = 0
+
+        _timer_noise_level = 0.05
+        _timer_cost = 100e-9  # 100 ns
+
+        _function_noise_level = 0.05
+        _function_costs = (
+            ("pass", 8e-9),
+            ("cheap_fn()", 4e-6),
+            ("expensive_fn()", 20e-6),
+        )
+
+        def __init__(self, stmt, setup, timer, globals):
+            self._random_state = np.random.RandomState(seed=self._seed)
+            self._mean_cost = {k: v for k, v in self._function_costs}[stmt]
+
+        def sample(self, mean, noise_level):
+            return max(self._random_state.normal(mean, mean * noise_level), 5e-9)
+
+        def timeit(self, number):
+            return sum([
+                # First timer invocation
+                self.sample(self._timer_cost, self._timer_noise_level),
+
+                # Stmt body
+                self.sample(self._mean_cost * number, self._function_noise_level),
+
+                # Second timer invocation
+                self.sample(self._timer_cost, self._timer_noise_level),
+            ])
+
+    def test_adaptive_timer(self):
+        class MockTimer(benchmark_utils.Timer):
+            _timer_cls = self._MockTimer
+
+        def assert_reprs_match(measurement, expected):
+            measurement_repr = re.sub(
+                "object at 0x[0-9a-fA-F]+>",
+                "object at 0xXXXXXXXXXXXX>",
+                repr(measurement)
+            )
+            self.assertEqual(measurement_repr, textwrap.dedent(expected).strip())
+
+        assert_reprs_match(
+            MockTimer("pass").blocked_autorange(min_run_time=10),
+            """
+            <torch.utils.benchmark.utils.common.Measurement object at 0xXXXXXXXXXXXX>
+            pass
+              Median: 7.98 ns
+              IQR:    0.52 ns (7.74 to 8.26)
+              125 measurements, 10000000 runs per measurement, 1 thread"""
+        )
```

Comment:
Wouldn't it be a lot better to use an expect test here? Then if you ever decide to change the formatting of how things print, you don't have to manually go through and update all of these strings. Expect test implementation lives in `torch.testing._internal.expecttest` and there are plenty of examples all around (a recent one from me is in `test/test_dispatch.py`

Reply:
Done.